Let RPC return FutureIValue instead of FutureMessage

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37536 Cleanup internal functions in python_functions.cpp
* #37520 Minor format cleanup in py_rref.cpp
* **#37519 Let RPC return FutureIValue instead of FutureMessage**

closes #37446

Currently FutureMessage is used in several places:

1. `rpc_async` returns a `FutureMessage` object and we expose it
   as `torch.distributed.rpc.Future`. From applications perspective,
   they are expecting a `py::object` instead of a `Message`, and we
   do the conversion in the `Future.wait()` pybind method.
2. RPC autograd profiler takes `FutureMessage` and installs
   callbacks to it. The profiler actually only need a `Future<T>`
   and does not care what `T` is.
3. `OwnerRRef` exposes a `getFuture()` API which returns a
   `FutureMessage`. This `FutureMessage` will be marked completed
   when the value referenced by the `OwnerRRef` is ready.
   `OwnerRRef` does not need it to be a Message type either, it
   actually creates an empty `Message` to mark the `Future`.

The above places are using `FutureMessage`, but they don't really
need a `Message`, and `Message` is a communication layer type that
applications or profiler or the RRef shouldn't be aware of.

Another motivation for making this change is that for async RPC
UDF #36071, we are going to allow application to call
`markCompleted` in Python. If we still use `FutureMessage`, then
in the `markCompleted` pybind function, it needs to convert the
provided `py::object` into a specific message type, which is
leaking communication layer code to pybind functions. Even if
this is doable, we will have two entities (RPC agent and pybind
Python frontend) accessing the same request callback logic. This is too messy.

This commit replaces all surface `FutureMessage` with `FutureIValue`,
so that `FutureMessage` is no longer visible from Python land. Note
that this does not cause BC issues, as the Python Future type name
and its API stay intact. Internally, we still have `FutureMessage`
in the communication layer.

Differential Revision: [D21308887](https://our.internmc.facebook.com/intern/diff/D21308887)

Commit:
Let RPC return FutureIValue instead of FutureMessage

closes #37446

Currently FutureMessage is used in several places:

1. `rpc_async` returns a `FutureMessage` object and we expose it
   as `torch.distributed.rpc.Future`. From applications perspective,
   they are expecting a `py::object` instead of a `Message`, and we
   do the conversion in the `Future.wait()` pybind method.
2. RPC autograd profiler takes `FutureMessage` and installs
   callbacks to it. The profiler actually only need a `Future<T>`
   and does not care what `T` is.
3. `OwnerRRef` exposes a `getFuture()` API which returns a
   `FutureMessage`. This `FutureMessage` will be marked completed
   when the value referenced by the `OwnerRRef` is ready.
   `OwnerRRef` does not need it to be a Message type either, it
   actually creates an empty `Message` to mark the `Future`.

The above places are using `FutureMessage`, but they don't really
need a `Message`, and `Message` is a communication layer type that
applications or profiler or the RRef shouldn't be aware of.

Another motivation for making this change is that for async RPC
UDF #36071, we are going to allow application to call
`markCompleted` in Python. If we still use `FutureMessage`, then
in the `markCompleted` pybind function, it needs to convert the
provided `py::object` into a specific message type, which is
leaking communication layer code to pybind functions. Even if
this is doable, we will have two entities (RPC agent and pybind
Python frontend) accessing the same request callback logic. This is too messy.

This commit replaces all surface `FutureMessage` with `FutureIValue`,
so that `FutureMessage` is no longer visible from Python land. Note
that this does not cause BC issues, as the Python Future type name
and its API stay intact. Internally, we still have `FutureMessage`
in the communication layer.

[ghstack-poisoned]

mrshenli committed:
Update on "Let RPC return FutureIValue instead of FutureMessage"


closes #37446

Currently FutureMessage is used in several places:

1. `rpc_async` returns a `FutureMessage` object and we expose it
   as `torch.distributed.rpc.Future`. From applications perspective,
   they are expecting a `py::object` instead of a `Message`, and we
   do the conversion in the `Future.wait()` pybind method.
2. RPC autograd profiler takes `FutureMessage` and installs
   callbacks to it. The profiler actually only need a `Future<T>`
   and does not care what `T` is.
3. `OwnerRRef` exposes a `getFuture()` API which returns a
   `FutureMessage`. This `FutureMessage` will be marked completed
   when the value referenced by the `OwnerRRef` is ready.
   `OwnerRRef` does not need it to be a Message type either, it
   actually creates an empty `Message` to mark the `Future`.

The above places are using `FutureMessage`, but they don't really
need a `Message`, and `Message` is a communication layer type that
applications or profiler or the RRef shouldn't be aware of.

Another motivation for making this change is that for async RPC
UDF #36071, we are going to allow application to call
`markCompleted` in Python. If we still use `FutureMessage`, then
in the `markCompleted` pybind function, it needs to convert the
provided `py::object` into a specific message type, which is
leaking communication layer code to pybind functions. Even if
this is doable, we will have two entities (RPC agent and pybind
Python frontend) accessing the same request callback logic. This is too messy.

This commit replaces all surface `FutureMessage` with `FutureIValue`,
so that `FutureMessage` is no longer visible from Python land. Note
that this does not cause BC issues, as the Python Future type name
and its API stay intact. Internally, we still have `FutureMessage`
in the communication layer.

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -22,6 +22,31 @@ namespace rpc {
+  delete fv;
+}
+
+std::shared_ptr<FutureIValue> toFutureIValue(
```
There are 2 definitions of `toFutureIValue()`, can we possibly dedupe these? Is there guidance on which API to use?

mrshenli replied:
Oh, the one in `python_function.cpp` assumes the IValue contains a `py::object`, and the one in `py_rref.cpp` does not, it is only for signaling the completion. Yes, let me consolidate these two into a util function.

rohan-varma replied:
If it's tricky to dedupe because of GIL, custom deleter, etc maybe rename such as `toFutureIValue` and `toFutureIValuePyObj` or something of the like

mrshenli replied:
This should be OK fo now, as both scenarios are Python-only (`python_function.cpp` and `py_rref.cpp`). Will check if this can be an issue when we merge this with `ivalue::Future`

rohan-varma commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -128,7 +128,8 @@ TORCH_API Message createExceptionResponse(const std::exception& e, int64_t id);
 
-typedef torch::utils::Future<Message> FutureMessage;
+using FutureMessage = torch::utils::Future<Message>;
+using FutureIValue = torch::utils::Future<at::IValue>;
```
Any reason we are sticking with `torch::utils::Future` instead of using `c10::ivalue::Future`? Not sure what the agreement here is on eventually merging these - the way I understood it is that `torch::utils::Future` should be the generic future used across all of PT, and then the existing `c10::ivalue::Future` is just a `torch::utils::Future` that holds an `IValue`. Is this still the plan (cc @xush6528 @wanchaol )

wanchaol replied:
@rohan-varma there's still bunch of things we need to do (i.e. make APIs consistent) before really adopting this plan. For the use of here, I think we could just use `c10::ivalue::Future` here as it's now just holding a IValue instead of a generic Message type.

mrshenli replied:
Will switch to `ivalue::Future` in a followup PR. I wanted to make this PR as short as possible, as this has impacts on several different places, e.g., rref/profiler/python-frontend.

xush6528 replied:
If we want to pass a ivalue future to a profiler jit operator that takes future, we do need to merge them.

For the Python RPC API, we can return a ivalue::Future instead, which is set by callback of Future<Message>. and kill the PyThonFuture wrapper in rpc/init.cpp. It requires a bit more code change. It's ok to do it in the next PR.

rohan-varma commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -13,6 +13,17 @@ namespace rpc {
+  auto fv = std::make_shared<FutureIValue>();
+
+  fm->addCallback(
+      [fv](const FutureMessage& fm) { fv->markCompleted(IValue()); });
```
Nit: we also don't propagate the error here, do we need to do that?

mrshenli replied:
It doesn't seem we need that for the current usage, as it seems that the returrned `FutureIValue` is only used by the profiler, which only uses the callback feature on the Future. But let me pass the error to the new Future as we are going to move this to a utility function.

rohan-varma commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -128,7 +128,8 @@ TORCH_API Message createExceptionResponse(const std::exception& e, int64_t id);
 createExceptionResponse(const std::string& exceptionStr, int64_t id);
 
-typedef torch::utils::Future<Message> FutureMessage;
+using FutureMessage = torch::utils::Future<Message>;
```
Nit: Add a note that this is an internal type and the future sent back to users should be Future<IValue>?

xush6528 commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -366,16 +366,17 @@ class TORCH_API OwnerRRef final : public RRef {
  private:
   friend class RRefContext;
 
+  // TODO: consolidate value_ into future_
```
I tried to do this in https://github.com/pytorch/pytorch/pull/32608

mrshenli replied:
I see, thanks for the reminder. Let me replace the TODO with a pointer to that abandoned PR then.

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -22,6 +22,31 @@ namespace rpc {
 
 namespace {
 
+void DeleteFutureIValue(FutureIValue* fv) {
```
Is there garentee this IValue is always PyObj? Can you add assert of if condition here?

Commit:
Update on "Let RPC return FutureIValue instead of FutureMessage"


closes #37446

Currently FutureMessage is used in several places:

1. `rpc_async` returns a `FutureMessage` object and we expose it
   as `torch.distributed.rpc.Future`. From applications perspective,
   they are expecting a `py::object` instead of a `Message`, and we
   do the conversion in the `Future.wait()` pybind method.
2. RPC autograd profiler takes `FutureMessage` and installs
   callbacks to it. The profiler actually only need a `Future<T>`
   and does not care what `T` is.
3. `OwnerRRef` exposes a `getFuture()` API which returns a
   `FutureMessage`. This `FutureMessage` will be marked completed
   when the value referenced by the `OwnerRRef` is ready.
   `OwnerRRef` does not need it to be a Message type either, it
   actually creates an empty `Message` to mark the `Future`.

The above places are using `FutureMessage`, but they don't really
need a `Message`, and `Message` is a communication layer type that
applications or profiler or the RRef shouldn't be aware of.

Another motivation for making this change is that for async RPC
UDF #36071, we are going to allow application to call
`markCompleted` in Python. If we still use `FutureMessage`, then
in the `markCompleted` pybind function, it needs to convert the
provided `py::object` into a specific message type, which is
leaking communication layer code to pybind functions. Even if
this is doable, we will have two entities (RPC agent and pybind
Python frontend) accessing the same request callback logic. This is too messy.

This commit replaces all surface `FutureMessage` with `FutureIValue`,
so that `FutureMessage` is no longer visible from Python land. Note
that this does not cause BC issues, as the Python Future type name
and its API stay intact. Internally, we still have `FutureMessage`
in the communication layer.

Differential Revision: [D21308887](https://our.internmc.facebook.com/intern/diff/D21308887)

[ghstack-poisoned]

mrshenli committed:
Update on "Let RPC return FutureIValue instead of FutureMessage"


closes #37446

Currently FutureMessage is used in several places:

1. `rpc_async` returns a `FutureMessage` object and we expose it
   as `torch.distributed.rpc.Future`. From applications perspective,
   they are expecting a `py::object` instead of a `Message`, and we
   do the conversion in the `Future.wait()` pybind method.
2. RPC autograd profiler takes `FutureMessage` and installs
   callbacks to it. The profiler actually only need a `Future<T>`
   and does not care what `T` is.
3. `OwnerRRef` exposes a `getFuture()` API which returns a
   `FutureMessage`. This `FutureMessage` will be marked completed
   when the value referenced by the `OwnerRRef` is ready.
   `OwnerRRef` does not need it to be a Message type either, it
   actually creates an empty `Message` to mark the `Future`.

The above places are using `FutureMessage`, but they don't really
need a `Message`, and `Message` is a communication layer type that
applications or profiler or the RRef shouldn't be aware of.

Another motivation for making this change is that for async RPC
UDF #36071, we are going to allow application to call
`markCompleted` in Python. If we still use `FutureMessage`, then
in the `markCompleted` pybind function, it needs to convert the
provided `py::object` into a specific message type, which is
leaking communication layer code to pybind functions. Even if
this is doable, we will have two entities (RPC agent and pybind
Python frontend) accessing the same request callback logic. This is too messy.

This commit replaces all surface `FutureMessage` with `FutureIValue`,
so that `FutureMessage` is no longer visible from Python land. Note
that this does not cause BC issues, as the Python Future type name
and its API stay intact. Internally, we still have `FutureMessage`
in the communication layer.

Differential Revision: [D21308887](https://our.internmc.facebook.com/intern/diff/D21308887)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -22,6 +22,38 @@ namespace rpc {
 
 namespace {
 
+py::object toPyObjInternal(RpcCommandBase& rpc, MessageType messageType) {
```
No change to these two functions, I just moved them from `torch::distributed::rpc` to an anonymous namespace, as `rpc/init.cpp` no longer needs to access this function.

mrshenli replied:
Let create a followup PR on top to merge these two functions.

mrshenli replied:
#37536
